src/process_quantiles.R

```diff
@@ -0,0 +1,184 @@
+# Processing code to create percentiles for groups of HRUs
+
+# This is preliminary work.
+
+# Creating a historic understanding of how "total storage"
+#   has varied through time for each HRU by calculating
+#   quantiles to establish very low, low, normal, high, 
+#   and very high thresholds of "total storage" for each
+#   HRU at every day of the year. Rather than only using
+#   values at every day of the year, we are using 5 values
+#   before and 5 values after to calculate the thresholds 
+#   for each day.
+
+# Total storage = pkwater_equiv + soil_moist_tot + hru_intcpstor +
+#                 hru_impervstor + gwres_stor + dprst_stor_hru
+
+# TO DO: 
+#   1. Use task_id to loop through groups of HRUs using slurm
+#   2. Calc for probs = seq(0, 1, by = 0.05)
+#   3. Fix this to work with incomplete years & leap years
+#   4. Add -Inf and +Inf in `combine_hru_quantiles` right before saving.
+
+start <- Sys.time()
+library(ncdf4) 
+library(data.table)
+
+########## Functions ########## 
+task_id_to_hru_seq <- function(task_id, n_hrus_per_task = 1000) {
+  # Convert task id to HRU ids (increment by n_hrus_per_task)
+  hru_id_start <- ((task_id-1)*n_hrus_per_task + 1) # which hru to start with
+  hru_id_end <- hru_id_start + (n_hrus_per_task-1) # which hru to end with
+  
+  # If we are past the last column, cut the final col id off
+  n_hrus <- 109951
+  hru_id_end <- ifelse(hru_id_end > n_hrus,
+                       yes = n_hrus,
+                       no = hru_id_end)
+  
+  hru_seq <- hru_id_start:hru_id_end
+  return(hru_seq)
+}
+read_ncdf_data <- function(fn, varid, hru_seq) {
+  
+  # The file is NetCDF
+  nc <- nc_open(fn)
+  
+  message(sprintf("Reading in NetCDF data for %s", varid))
+  # Only load rows for current HRUs
+  data_nc <- ncvar_get(nc, varid, 
+                       start = c(head(hru_seq, 1), 1),
+                       count = c(length(hru_seq), -1))
+  time <- ncvar_get(nc, "time")
+  hruid <- ncvar_get(nc, "hruid")[hru_seq]
+  
+  # Convert ncdf4 times to R dates
+  time_att <- ncdf4::ncatt_get(nc, "time")
+  time_start <- as.Date(gsub("days since ", "", time_att$units))
+  time_fixed <- time + time_start # creates dates from the "days since" var
+  
+  # Keep only complete years of data
+  # data_years <- format(time_fixed, "%Y")
+  # n_days_per_year <- table(data_years)
+  # data_years_complete <- names(n_days_per_year)[n_days_per_year >= 365]
+  
+  year_doy_vector <- paste(format(time_fixed, "%Y"), format(time_fixed, "%j"), sep = "_")
+  dt <- as.data.table(cbind(hruid, data_nc)) # takes ~2 minutes
```

Comment:
Now my computer is not recreating the 2 min issue, but it was taking awhile. Since the HRUIDs are really just the column numbers, I might be able to skip the `cbind` altogether.